api/src/main/java/org/openmrs/api/impl/CohortServiceImpl.java

```diff
@@ -57,108 +59,160 @@ public Cohort saveCohort(Cohort cohort) throws APIException {
 		return dao.getCohortsContainingPatientId(patientId);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohorts(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getCohorts(String nameFragment) throws APIException {
 		return dao.getCohorts(nameFragment);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts()
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts() throws APIException {
 		return Context.getCohortService().getAllCohorts(false);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getAllCohorts(boolean)
 	 */
 	@Transactional(readOnly = true)
 	public List<Cohort> getAllCohorts(boolean includeVoided) throws APIException {
 		return dao.getAllCohorts(includeVoided);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohort(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public Cohort getCohort(String name) throws APIException {
 		return dao.getCohort(name);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#purgeCohort(org.openmrs.Cohort)
 	 */
 	public Cohort purgeCohort(Cohort cohort) throws APIException {
 		return dao.deleteCohort(cohort);
 	}
+
+	/**
+	 * @see org.openmrs.api.CohortService#addMembershipToCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort addMembershipToCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+        cohort.addMembership(cohortMembership);
+		Context.getCohortService().saveCohort(cohort);
+		return cohort;
+	}
+
+	/**
+	 * @see org.openmrs.api.CohortService#removeMemberShipFromCohort(org.openmrs.Cohort,
+	 * 		org.openmrs.CohortMembership)
+	 */
+	public Cohort removeMemberShipFromCohort(Cohort cohort, CohortMembership cohortMembership) throws APIException {
+		if (cohort.contains(cohortMembership.getPatient())) {
+			cohort.removeMembership(cohortMembership);
+			Context.getCohortService().saveCohort(cohort);
+		}
+		return cohort;
+	}
+
+	/**
+     * @see org.openmrs.api.CohortService#patientVoided(org.openmrs.Patient)
+	 */
+	public void patientVoided(Patient patient) throws APIException {
+        List<Cohort> cohorts = Context.getCohortService().getCohortsContainingPatient(patient);
+        for (Cohort cohort : cohorts) {
+			cohort.removeMember(patient, "Patient associated to member is voided");
+			Context.getCohortService().saveCohort(cohort);
+		}
+	}
 	
+	/**
+	 * @see org.openmrs.api.CohortService#patientUnvoided(org.openmrs.Patient)
+	 */
+	public void patientUnvoided(Patient patient) throws APIException {
+		List<Cohort> cohorts = Context.getCohortService().getCohortsContainingPatient(patient);
+		for (Cohort cohort : cohorts) {
+			List<CohortMembership> membersToUnvoid = cohort.getMembers().stream()
+					.filter(m -> m.getVoided() && m.getPatient().getPatientId().equals(patient.getPatientId()))
+					.collect(Collectors.toList());
+			for (CohortMembership member : membersToUnvoid) {
```

Comment:
I understand, but I didn't get my question answered. I'll explain it clearly:

PatientDataUnvoidHandler.java calls CohortService.patientUnvoided(Patient), at this point before the call is made, is the Patient unvoided i.e. are the void fields unset? In short, what is the state of the patient before patientUnvoided(Patient) is called? I was assuming the patients were unvoided with their void fields unset.

If the patient is still voided, and patientUnvoided(Patient) is called, won't I have to unvoid the patient as well?